neqo-http3/src/push_stream.rs

```diff
@@ -0,0 +1,162 @@
+// `PushController` that will set the push state to closed and remove any push events.
+
+#[derive(Debug)]
+pub(crate) struct PushStream {
+    state: PushStreamState,
+    stream_id: u64,
+    push_handler: Rc<RefCell<PushController>>,
+    events: Http3ClientEvents,
+}
+
+impl PushStream {
+    pub fn new(
+        stream_id: u64,
+        push_handler: Rc<RefCell<PushController>>,
+        events: Http3ClientEvents,
+    ) -> Self {
+        Self {
+            state: PushStreamState::ReadPushId(NewStreamTypeReader::new()),
+            stream_id,
+            push_handler,
+            events,
+        }
+    }
+}
+
+impl Display for PushStream {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        write!(f, "Push stream {:?}", self.stream_id)
+    }
+}
+
+impl RecvStream for PushStream {
+    fn receive(&mut self, conn: &mut Connection, decoder: &mut QPackDecoder) -> Res<()> {
+        loop {
+            match &mut self.state {
+                PushStreamState::ReadPushId(id_reader) => {
+                    let push_id = id_reader.get_type(conn, self.stream_id);
+                    let fin = id_reader.fin();
+                    if fin {
+                        self.state = PushStreamState::Closed;
+                        return Ok(());
+                    }
+                    if let Some(p) = push_id {
+                        if self
+                            .push_handler
+                            .borrow_mut()
+                            .add_new_push_stream(p, self.stream_id)?
+                        {
+                            self.state = PushStreamState::ReadResponse {
+                                push_id: p,
+                                response: RecvMessage::new(
+                                    self.stream_id,
+                                    Box::new(RecvPushEvents::new(p, self.push_handler.clone())),
+                                    None,
+                                ),
+                            };
+                        } else {
+                            let _ = conn.stream_stop_sending(
+                                self.stream_id,
+                                Error::HttpRequestCancelled.code(),
+                            );
+                            self.state = PushStreamState::Closed;
+                            return Ok(());
+                        }
+                    }
+                }
+                PushStreamState::ReadResponse { response, push_id } => {
+                    response.receive(conn, decoder)?;
+                    if response.done() {
+                        self.push_handler.borrow_mut().close(*push_id);
+                        self.state = PushStreamState::Closed;
+                    }
+                    return Ok(());
+                }
+                _ => return Ok(()),
+            }
+        }
+    }
+
+    fn done(&self) -> bool {
+        matches!(self.state, PushStreamState::Closed)
```

Comment:
It's probably better to use `==` rather than `matches!()` unless you really have to.

Reply:
Implementing PartalEq for PushStreamState is not easy (NewStreamTypeReader, RecvMessage, HFrameReader, etc need to implement PartilaEq).
I could do a partial implementation of it, but I decided not to do it